src/components/core/Dropdown/Dropdown.tsx

```diff
@@ -0,0 +1,107 @@
+import {
+  ReactNode,
+  RefObject,
+  useCallback,
+  useEffect,
+  useRef,
+  useState,
+} from 'react';
+import styled from 'styled-components';
+import TextButton from '../Button/TextButton';
+import colors from '../colors';
+
+type Props = {
+  mainText: string;
+  children?: ReactNode;
+};
+
+const refContainsEventTarget = (ref: RefObject<HTMLDivElement>, event: any) => {
+  return ref.current && ref.current.contains(event.target);
+};
+
+function Dropdown({ mainText, children }: Props) {
+  const [isDropdownVisible, setIsDropdownVisible] = useState(false);
+
+  const handleClick = useCallback(() => {
+    setIsDropdownVisible(!isDropdownVisible);
+  }, [isDropdownVisible]);
+
+  const dropdownMenuRef = useRef<HTMLDivElement>(null);
+  const dropdownButtonRef = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
+    function handleDocumentClick(event: any) {
```

Comment:
i think this function should be defined as a `useCallback`, and we should pass an argument into the function about whether it's open (e.g. `handleDocumentClick(event: any, isDropdownVisible: boolean)`. this will make it so the callback doesn't have to depend on any react state, and it'll be a "pure function"

we want a stable function in memory so that `document.addEventListener(x, function)` will register that function once and only once, even if called multiple times.

example:
```js
// adds 3 event listners
document.addEventListener('mousedown', () => console.log('hi'))
document.addEventListener('mousedown', () => console.log('hi'))
document.addEventListener('mousedown', () => console.log('hi'))

// adds only 1 event listener!
const sayHi = () => console.log('hi')
document.addEventListener('mousedown', sayHi)
document.addEventListener('mousedown', sayHi)
document.addEventListener('mousedown', sayHi)
```

so basically within our effect if we don't want to worry about setting `handleDocumentClick` listener too many times, it should be a callback defined outside of the `useEffect` that's defined only once

Reply:
good all- I left a more detailed reply below